pkg/api/config.go

```diff
@@ -1,27 +1,199 @@
 package api
 
-import "errors"
+import (
+	"fmt"
+	"net/url"
+	"reflect"
+	"sort"
+	"strings"
 
-// Validate validates config
-func (config *ReleaseBuildConfiguration) Validate() error {
-	buildRootImage := config.InputConfiguration.BuildRootImage
-	testBaseImage := config.InputConfiguration.TestBaseImage
-	if config.Tests != nil {
-		for _, test := range config.Tests {
-			if test.As == "images" {
-				return errors.New("test should not be called 'images' because it gets confused with '[images]' target")
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+var (
+	inputConfigurationType          = reflect.TypeOf(InputConfiguration{})
+	buildRootImageConfigurationType = reflect.TypeOf(BuildRootImageConfiguration{})
+	imageStreamTagReferenceType     = reflect.TypeOf(ImageStreamTagReference{})
+	testStepConfigurationType       = reflect.TypeOf(TestStepConfiguration{})
+	imageStreamTagReferenceMapType  = reflect.TypeOf(map[string]ImageStreamTagReference{})
+)
+
+func deepValidate(iface reflect.Value, fieldRoot []reflect.StructTag) error {
+	var validationErrors []error
+	if reflect.Zero(iface.Type()) == iface {
+		return nil
+	}
+
+	// Resolve pointers value.
+	if iface.Kind() == reflect.Ptr {
+		iface = reflect.Indirect(iface)
+	}
+
+	switch iface.Type() {
+	// Validate InputConfiguration
+	case inputConfigurationType:
+		validationErrors = append(validationErrors,
+			validateInputConfiguration(fieldRoot, iface.Interface().(InputConfiguration)))
+
+	// Validate BuildRootImageConfiguration
+	case buildRootImageConfigurationType:
+		validationErrors = append(validationErrors,
+			validateBuildRootImageConfiguration(fieldRoot, iface.Interface().(BuildRootImageConfiguration)))
+
+	// Validate ImageStreamTagReference
+	case imageStreamTagReferenceType:
+		validationErrors = append(validationErrors,
+			validateImageStreamTagReference(fieldRoot, iface.Interface().(ImageStreamTagReference)))
+	}
+
+	if iface.Kind() == reflect.Struct {
```

Comment:
why are we doing reflection here?  I don't see how reflection makes this better at validation (for one thing, I can't look to see whether you are checking structs).